oauth/scope.js

```diff
@@ -0,0 +1,212 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+const { URL } = require('url');
+
+// These character ranges are from the OAuth RFC,
+// https://tools.ietf.org/html/rfc6749#section-3.3
+const VALID_SCOPE_VALUE = /^[\x21\x23-\x5B\x5D-\x7E]+$/;
+
+const VALID_SHORT_NAME_VALUE = /^[a-zA-Z0-9_]+$/;
+const VALID_FRAGMENT_VALUE = /^#[a-zA-Z0-9_]+$/;
+
+
+class ShortScopeValue {
+
+  constructor(value) {
+    if (! VALID_SCOPE_VALUE.test(value)) {
+      throw new Error('Invalid scope value: ' + value);
+    }
+    this.hasWrite = false;
+    this.names = value.split(':');
+    this.names.forEach(name => {
+      if (! VALID_SHORT_NAME_VALUE.test(name)) {
+        throw new Error('Invalid scope value: ' + value);
+      }
+    })
+    if (this.names[this.names.length - 1] === 'write') {
+      this.hasWrite = true;
+      this.names.pop();
+      if (this.names.length === 0) {
+        throw new Error('Invalid scope value: ' + value);
+      }
+    }
+  }
+
+  toString() {
+    let s = this.names.join(':');
+    if (this.hasWrite) {
+      s += ':write';
+    }
+    return s
+  }
+
+  implies(value) {
+    if (! value.names) {
+      return false;
+    }
+    if (value.hasWrite) {
+      if (! this.hasWrite) {
+        return false;
+      }
+    }
+    if (value.names.length < this.names.length) {
+      return false;
+    }
+    for (let i = 0; i < this.names.length; i++) {
+      if (value.names[i] !== this.names[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+
+class URLScopeValue {
+
+  constructor(value) {
+    if (! VALID_SCOPE_VALUE.test(value)) {
+      throw new Error('Invalid scope value: ' + value);
+    }
+    this.url = new URL(value);
+    if (this.url.protocol !== 'https:') {
+      throw new Error('Invalid scope value: ' + value);
+    }
+    if (this.url.username || this.url.password || this.url.query) {
```

Comment:
Scope URLs are not allowed to have usernames, passwords, or query params?

Reply:
That's correct; I don't see why they'd need them